crates/data-generator/src/blocks/json.rs

```diff
@@ -0,0 +1,153 @@
+            Kind::Numeric => {
+                let vars = prop_vars.into_iter().map(|s| s.parse().unwrap()).collect();
+                PropertyVariants::Numeric(vars)
+            }
+            Kind::Enum => {
+                let vars = prop_vars.into_iter().map(Name::from_snake_case).collect();
+                PropertyVariants::Enum(vars)
+            }
+        }
+    }
+
+    fn state(
+        all_properties: &IndexMap<Name, PropertyVariants>,
+        state_id: u32,
+        state_props: BTreeMap<String, String>,
+    ) -> (Id, State) {
+        let properties = state_props
+            .into_iter()
+            .map(|(name_raw, variant_raw)| {
+                let prop_name = Name::from_snake_case(name_raw);
+
+                let variants = all_properties
+                    .iter()
+                    .find_map(|(name, vars)| (name == &prop_name).then_some(vars))
+                    .unwrap();
+
+                let prop_var = PropertyVariant::from(variant_raw, variants);
+                (prop_name, prop_var)
+            })
+            .collect();
+
+        let id = Id(state_id);
+        let block_state = State { properties };
+        (id, block_state)
+    }
+
+    let result = registry
+        .into_iter()
+        .map(|(block_name_raw, serde_block)| {
+            let block_name = consume_until_colon(&block_name_raw);
+            let block_name = Name::from_snake_case(block_name);
+
+            let block_props = serde_block
+                .properties
+                .into_iter()
+                .map(|(prop_raw, vars_raw)| {
+                    let name = Name::from_snake_case(prop_raw);
+                    let vars = property_variants(vars_raw);
+                    (name, vars)
+                })
+                .collect();
+
+            let mut block_states = IndexMap::with_capacity(serde_block.states.len());
+            let mut def_state_id = Id(0);
+            for s in serde_block.states.into_iter() {
+                if s.default {
+                    def_state_id = Id(s.id.0);
+                }
+                let (k, v) = state(&block_props, s.id.0, s.properties);
+                block_states.insert(k, v);
+            }
+
+            let block = Block {
+                properties: block_props,
+                states: States {
+                    states: block_states,
+                    default: def_state_id,
+                },
+            };
+
+            (block_name, block)
+        })
+        .collect();
+
+    Ok(result)
+}
+
+fn consume_until_colon(input: &'_ str) -> &'_ str {
+    &input[input.find(':').map_or(0, |pos| pos + 1)..]
+}
+
+fn all_nums(variants: impl IntoIterator<Item = impl AsRef<str>>) -> bool {
+    variants
+        .into_iter()
+        .all(|s| s.as_ref().chars().all(|c| c.is_numeric()))
+}
+
+fn all_bools<'a, I, II>(variants: II) -> bool
+where
+    I: AsRef<str> + ?Sized + 'a,
+    II: IntoIterator<Item = &'a I>,
+{
```

Comment:
This is just hard to read, besides we're not in library code
```suggestion
fn all_bools<'a>(variants: impl IntoIterator<Item = &'a str>) -> bool {
```